File: AbercrombieFitchCodeTest/app/src/main/java/com/travelapp/abercrombiefitchcodetest/MainActivity.java
package com.travelapp.abercrombiefitchcodetest;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();
    private RecyclerView recyclerView;
    private RecyclerView.LayoutManager layoutManager;
    private RecyclerViewAdapter recyclerViewAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ApiInterface apiInterface = RetrofitInstance.getRetrofitInstance().create(ApiInterface.class);
        recyclerView = findViewById(R.id.recyclerview);
        layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);

        Observable<List<ShoppingModel>> shoppingModelObservable = apiInterface.getShopping();
        shoppingModelObservable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::handleResult, this::handleError);
    }

    private void handleResult(List<ShoppingModel> shoppingModels) {
        Log.d(TAG, "handleResult: " + shoppingModels);
        recyclerViewAdapter = new RecyclerViewAdapter(shoppingModels, MainActivity.this);
        recyclerView.setAdapter(recyclerViewAdapter);
    }

    private void handleError(Throwable throwable) {
        Log.d(TAG, "handleError: " +  throwable.getMessage());
    }

}

File: KotlinRetrofitRxJavaRecyclerViewTough/app/src/main/java/com/calvert/kotlinretrofitrxjavarecyclerviewtough/MainActivity.kt
package com.calvert.kotlinretrofitrxjavarecyclerviewtough

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import com.calvert.kotlinretrofitrxjavarecyclerviewtough.adapter.MyAdapter
import com.calvert.kotlinretrofitrxjavarecyclerviewtough.model.User
import com.calvert.kotlinretrofitrxjavarecyclerviewtough.network.ApiInterface
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    lateinit var adapter: MyAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        recyclerView.layoutManager = LinearLayoutManager(this)

        val apiInterface = ApiInterface.getRetrofit().getUsers()
//        apiInterface.enqueue(object: Callback<User>{
//            override fun onFailure(call: Call<User>, t: Throwable) {
//                Log.e("Error", t.message)
//            }
//
//            override fun onResponse(call: Call<User>, response: Response<User>) {
//                Log.d("Response", response.body().toString())
//                adapter = MyAdapter(this@MainActivity, response.body()?.data!!.toList())
//                recyclerView.adapter = adapter
//                adapter.notifyDataSetChanged()
//            }
//        })
            // do in background thread
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({it ->
                Log.d("response", it.toString())
                adapter = MyAdapter(this@MainActivity, it?.data!!.toList())
                recyclerView.adapter = adapter
            }, {it ->
                Log.e("failure", it.message)
            })
    }
}
